Implement liminal sdk event manager

PSDIUM opened the pull request:
Implement a static script to allow actions to be queued on both platform and sdk side that can be then triggered by either side when required.

PSDIUM committed:
Implement liminal sdk event manager

Implement a static script to allow actions to be queued on both platform and sdk side that can be then triggered by either side when required.

tinnystudios commented on SDK/Systems/EventManager/LiminalSdkEventManager.cs:
```diff
@@ -0,0 +1,57 @@
+    public static class LiminalSdkEventManager
+    {
+
+        private static Queue<Action> _queuedPlatformActions = new Queue<Action>();
```
Action is fine but this means it's almost always firing immediately. 

For scalability I would suggest actually have it as a class such as QueueActionEventData or QueueActionModel containing the Action. 

Down the track you can add more features to it such as timing or any conditions.

tinnystudios commented on SDK/Systems/EventManager/LiminalSdkEventManager.cs:
```diff
@@ -0,0 +1,57 @@
+        /// Queues an event from the experience / liminal SDK side that can be triggered by the platform when necessary
+        /// </summary>
+        /// <param name="evnt"></param>
+        public static void QueueEventForPlatform(Action evnt)
```
Ah you wrote evnt because event is a keyword? 

I wouldn't use evnt here.

tinnystudios reviewed (changes requested):
I've added a few comments. 

I would also suggest adding an example usage so the APIs are all worked out in 'practice'

PSDIUM committed:
Modify liminal sdk event manager to use a custom data class for events

replaced action queue for a list of custom class 'ActionEventData' which houses the action as well as additional properties. Added additional methods for calling individual actions as well as multiple actions at once. Converted methods into IEnumerators to allow for delayed actions and to make it more scalable in future.

PSDIUM committed:
Update QueueEvent method to AddEvent in sdk event manager

Updated the QueveEvent method to AddEvent to be more inline with how the system currently works.

PSDIUM committed:
Add sdk event manager example and adjust event manager to allow for events to be kept after use, and remove event methods

Added an example script of how to use the sdk event manager to register and trigger events. Updated the event manager to have functionality to keep events after they have been invoked for repeat use if necessary.  Added methods to remove action event data by id, allowing events to be removed when required.

PSDIUM committed:
Clear up typo in sdk event manager example script

tinnystudios commented on SDK/Systems/EventManager/LiminalSdkEventManager.cs:
```diff
@@ -0,0 +1,156 @@
+
+namespace Liminal.Shared
+{
+    public static class LiminalSdkEventManager
```
Id add a summary to the class too. Generally I don't mind self explanatory methods without summary but classes would be great!

tinnystudios replied:
I'd suggest a namespace more like this now because we have a data types in the same system. maybe Liminal.SDK.EventManager?

tinnystudios commented on SDK/Systems/EventManager/LiminalSdkEventManager.cs:
```diff
@@ -0,0 +1,156 @@
+        /// Remove all ActionEventData from experienceActionEventList with matching id
+        /// </summary>
+        /// <param name="id">Id of the ActionEventData</param>
+        public static void RemoveAllExperienceEventDataWithId(string id)
```
Love this!

tinnystudios commented on SDK/Systems/EventManager/LiminalSdkEventManager.cs:
```diff
@@ -0,0 +1,156 @@
+            list.RemoveAll(x => x.Id.Equals(id));
+        }
+
+        private static IEnumerator TriggerEvent(ActionEventData eventData)
```
I'd add some kind of tag to the Coroutine. 

TriggerEventRoutine();

tinnystudios commented on SDK/Systems/EventManager/LiminalSdkEventManager.cs:
```diff
@@ -0,0 +1,156 @@
+            eventData.Action?.Invoke();
+        }
+
+        public class ActionEventData
```
I'd move this out of the class into its own file. 

In this case, maybe update both to be under a more better namespace like Liminal.SDK.EventManager

tinnystudios commented on SDK/Systems/Examples/LiminalSdkActionEventExample.cs:
```diff
@@ -0,0 +1,39 @@
+using Liminal.Shared;
+using UnityEngine;
+
+public class LiminalSdkActionEventExample : MonoBehaviour
```
Namespace this, 

Liminal.SDK.EventManager.Examples

tinnystudios reviewed (changes requested):
Few more minor things.

PSDIUM committed:
Update liminal sdk event manager to separate out event data class, rename methods, change namespace

Made adjustments to the sdk event manager to move event data class into its own class rather than be nested, renamed coroutine methods to include the 'routine' convention in the naming, and adjusted the namespace to something more appropriate.

tinnystudios commented:
Looks good now! Good work!
